### PR TITLE
Adding new functions for templates

### PR DESCRIPTION
**Pull request by dockerymick**
https://issues.redhat.com/browse/ACM-3116

### CODE REVIEW

**Comment by dockerymick** [comment]
@gparvin @ckandag thanks for your time and feedback during the live review. i have made some more updates. When you have an opportunity, will you both review it again?

File: governance/adv_template_process.adoc [hunk 1]
```diff
@@ -0,0 +1,35 @@
+[#adv-template-processing]
+= Advanced template processing in configuration policies
+
+Use both managed cluster and hub cluster templates to reduce the need to create separate policies for each target cluster or hardcode configuration values in the policy definitions. For security, both resource-specific and the generic lookup functions in hub cluster templates are restricted to the namespace of the policy on the hub cluster.
+
+*Important:* If you use hub cluster templates to propagate secrets or other sensitive data, the sensitive data exists in the managed cluster namespace on the hub cluster and on the managed clusters where that policy is distributed. The template content is expanded in the policy, and policies are not encrypted by the {ocp-short} ETCD encryption support. To address this, use `fromSecret`, which automatically encrypts the values from the Secret, or `protect` to encrypt other values.
```
Comment (oafischer) [inline governance/adv_template_process.adoc]: Does this mean that if you use hub cluster templates to propogate secrets, that causes the sensitive data to exist in the managed cluster namespace, or is the data there anyway and we are just informing the user where they can find it if they are using hub cluster templates?

File: governance/adv_template_process.adoc
Comment (oafischer) [inline governance/adv_template_process.adoc:1]: Good file names!

File: governance/template_functions.adoc [hunk 2]
```diff
@@ -1,29 +1,9 @@
...
+= Template functions
 
-To conform templates with YAML syntax, templates must be set in the policy resource as strings using quotes or a block character (`|` or `>`). This causes the resolved template value to also be a string. To override this, consider using `toInt` or `toBool` as the final function in the template to initiate further processing that forces the value to be interpreted as an integer or boolean respectively.
+Template functions, such as resource-specific and generic `lookup` template functions, are available for referencing Kubernetes resources on the hub cluster (using the `{{hub ... hub}}` delimiters), or managed cluster (using the `{{ ... }}` delimiters). See _Support for hub cluster templates in configuration policies_ for more details. The resource-specific functions are used for convenience and makes content of the resources more accessible. If you use the generic function, `lookup`, which is more advanced, it is best to be familiar with the YAML structure of the resource that is being looked up. In addition to these functions, utility functions like `base64encode`, `base64decode`, `indent`, `autoindent`, `toInt`, `toBool`, and more are also available.
```
Comment (oafischer) [inline governance/template_functions.adoc]:
```suggestion
Template functions, such as resource-specific and generic `lookup` template functions, are available for referencing Kubernetes resources on the hub cluster (using the `{{hub ... hub}}` delimiters), or managed cluster (using the `{{ ... }}` delimiters). See _Support for hub cluster templates in configuration policies_ for more details. The resource-specific functions are used for convenience and makes content of the resources more accessible. If you use the generic function, `lookup`, which is more advanced, it is best to be familiar with the YAML structure of the resource that is being looked up. In addition to these functions, utility functions such as `base64encode`, `base64decode`, `indent`, `autoindent`, `toInt`, `toBool`, and more are also available.
```

File: governance/template_functions.adoc [hunk 3]
```diff
@@ -1,29 +1,9 @@
...
 
-Continue reading to view descriptions and examples for some of the custom template functions that are supported:
+To conform templates with YAML syntax, templates must be set in the policy resource as strings using quotes or a block character (`|` or `>`). This causes the resolved template value to also be a string. To override this, consider using `toInt` or `toBool` as the final function in the template to initiate further processing that forces the value to be interpreted as an integer or boolean respectively. Continue reading to view descriptions and examples for some of the custom template functions that are supported:
```
Comment (oafischer) [inline governance/template_functions.adoc]: "consider using" is a little vague, I think we can just leave "consider" out.

File: governance/template_functions.adoc [hunk 4]
```diff
@@ -452,8 +435,42 @@ After using the `toLiteral` function, the following update is displayed:
...
+----
+complianceType: musthave
+      objectDefinition:
+        apiVersion: v1
+        kind: ConfigMap
+        metadata:
+          name: my-secret-copy
+        data: '{{ fromConfigMapData "default" "my-configmap" }}'
+
+
```
Comment (oafischer) [inline governance/template_functions.adoc]: Looks like some empty space snuck in here

File: governance/adv_template_process.adoc (hunk 1 above)
Comment (dockerymick) [reply governance/adv_template_process.adoc]: @oafischer great question. After rereading it does seem like the propagation is the cause for sensitive data to be exposed in the manage cluster namespace.

File: governance/adv_template_process.adoc [hunk 5]
```diff
@@ -0,0 +1,35 @@
+[#adv-template-processing]
+= Advanced template processing in configuration policies
+
+Use both managed cluster and hub cluster templates to reduce the need to create separate policies for each target cluster or hardcode configuration values in the policy definitions. For security, both resource-specific and the generic lookup functions in hub cluster templates are restricted to the namespace of the policy on the hub cluster.
+
+*Important:* If you use hub cluster templates to propagate secrets or other sensitive data, that causes sensitive data exposure in the managed cluster namespace on the hub cluster and on the managed clusters where that policy is distributed. The template content is expanded in the policy, and policies are not encrypted by the {ocp-short} ETCD encryption support. To address this, use `fromSecret`, which automatically encrypts the values from the Secret, or `protect` to encrypt other values.
```
Comment (gparvin) [inline governance/adv_template_process.adoc]: Update to: use `fromSecret` or `copySecretData`,

Comment (dockerymick) [reply governance/adv_template_process.adoc]: Thanks for your patience Gus! Updated

File: governance/template_functions.adoc (hunk 4 above)
Comment (dockerymick) [reply governance/template_functions.adoc]: Updated

File: governance/template_functions.adoc (hunk 3 above)
Comment (dockerymick) [reply governance/template_functions.adoc]: Updated

**Comment by dockerymick** [comment]
@ckandag  @oafischer @gparvin thanks for your help and reviews. Will you all rereview the PR please?

**Review by oafischer — APPROVED** [review]
Looks great!

File: governance/adv_template_process.adoc (hunk 5 above)
Comment (dockerymick) [reply governance/adv_template_process.adoc]: Updated

**Comment by dockerymick** [comment]
@oafischer @gparvin thanks for the reviews and approvals Screenshot 2023-04-05 at 11 57 47 AM

Merging this PR

**Pull request merged**